035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+> It's also clear that there is a `Secret` which stores CA certificates, and their current state, and which needs to be accessed by the issuing operator and the cluster operator.
+
+#### Process for ensuring CA trust
+
+When an EE certificate is issued via a `CertificateIssuer` it might have been signed by a CA that is not yet trusted.
+The cluster operator (for the cluster CA), or the user operator (for the clients CA) needs to detect this lack of established trust and ask the cluster operator to trust the new CA. 
+The cluster operator and the user operator run in separate processes, so this will be done via a shared `Secret`, for example:
+
+2. The issuing operator has placed a root CA certificate in the shared `Secret` in the `UNTRUSTED` state.
+3. The cluster operator sees the root CA certificate orchestrates changes to the brokers (reconfig or restart) (and, for the Cluster CA the other components) so that the certificate is trusted by all components that will observe the EE certificates.
+4. The cluster operator updates the shared `Secret` to indicate that the certificate is now in the `TRUSTED_NEW` state.
+
+#### Process for deploying the issued certificate
+
+1. The issuing operator sees that the root CA certificate is now in the `TRUSTED_NEW` state, so can infer that the CA certificate is trusted.
+2. The issuing operator copies the EE certificate and key, and the rest of the certificate chain from its `Secret` _B_  to the deployment secret (e.g. the "user secret" for the Clients CA, or the relevant component secret for the Cluster CA).
+3. The issuing operator updates the state of the root CA certificate in the shared `Secret` to `TRUSTED_USED`. 
+
+#### Need for cert path validation
+
+The issuing operator needs to know when to request a new certificate for the same subject because the existing certificate is (or is soon going to be) invalid.
+Validity is a property of the whole certificate chain, not just the EE certificate.
+For example it is possible that a root or intermediate certificate expires before an EE certificate that depends on it. 
+So for the operator to reason correctly about validity it needs to perform PKIX cert path validation.
+Fortunately Java comes with APIs for doing this in `java.security.cert`.
+
+It is *not* proposed to integrate support for CRL or OSCP under this proposal, though it could be done in the future. 
+This is no different from the current situation in Strimzi. 
+
+####  EE cert expiry
+
+EE certificate expiry is driven by the cert path validation described above.
+The issuing operator will perform cert path validation and if the end entity certificate is currently expired, or will be expired on `now + renewal period` then a new EE certificate will be issued as described in the previous section.
+
+#### Intermediate or root CA cert expiry
```

Comment:
I think this deserves a bit more detail on how would it work. For example in my past investigation, Cert Manager didn't really allow a smooth migration from old CA to new CA because it just replaced the CA. That makes this a bit complicated and that is also why as mentioned in one of the comments above you might need a clear separartion of the secrets.

Reply:
Yes, I plan to update with which Secretss are updated in which steps.